src/progress.rs

```diff
@@ -841,4 +1149,101 @@ mod test_progress_set {
+        )
+    }
+
+    #[test]
+    fn test_membership_change_configuration_empty_sets() {
+        assert!(check_membership_change_configuration((vec![], vec![]), (vec![], vec![])).is_err())
+    }
+
+    #[test]
+    fn test_membership_change_configuration_empty_voters() {
+        assert!(
+            check_membership_change_configuration((vec![1], vec![]), (vec![], vec![]),).is_err()
+        )
+    }
+
+    #[test]
+    fn test_membership_change_configuration_add_voter() -> Result<()> {
+        check_membership_change_configuration((vec![1], vec![]), (vec![1, 2], vec![]))
+    }
+
+    #[test]
+    fn test_membership_change_configuration_add_learner() -> Result<()> {
+        check_membership_change_configuration((vec![1], vec![]), (vec![1], vec![2]))
+    }
+
+    #[test]
+    fn test_membership_change_configuration_promote_learner() -> Result<()> {
+        check_membership_change_configuration((vec![1], vec![2]), (vec![1, 2], vec![]))
+    }
+
+    fn check_membership_change_configuration(
+        start: (impl IntoIterator<Item = u64>, impl IntoIterator<Item = u64>),
+        end: (impl IntoIterator<Item = u64>, impl IntoIterator<Item = u64>),
+    ) -> Result<()> {
+        let start_voters = start.0.into_iter().collect::<FxHashSet<u64>>();
+        let start_learners = start.1.into_iter().collect::<FxHashSet<u64>>();
+        let end_voters = end.0.into_iter().collect::<FxHashSet<u64>>();
+        let end_learners = end.1.into_iter().collect::<FxHashSet<u64>>();
+        let transition_voters = start_voters
+            .union(&end_voters)
+            .cloned()
+            .collect::<FxHashSet<u64>>();
+        let transition_learners = start_learners
+            .union(&end_learners)
+            .cloned()
+            .collect::<FxHashSet<u64>>();
+
+        let mut set = ProgressSet::default();
+        let default_progress = Progress::new(0, 10);
+
+        for starter in start_voters {
+            set.insert_voter(starter, default_progress.clone())?;
+        }
+        for starter in start_learners {
+            set.insert_learner(starter, default_progress.clone())?;
+        }
+        set.begin_membership_change(
+            Configuration::new(end_voters.clone(), end_learners.clone()),
+            default_progress,
+        )?;
+        assert!(set.is_in_membership_change());
+        assert_eq!(
+            set.voter_ids(),
+            transition_voters,
+            "Transition state voters inaccurate"
+        );
+        assert_eq!(
+            set.learner_ids(),
+            transition_learners,
+            "Transition state learners inaccurate."
+        );
+
+        set.finalize_membership_change()?;
+        assert!(!set.is_in_membership_change());
+        assert_eq!(set.voter_ids(), end_voters, "End state voters inaccurate");
+        assert_eq!(
+            set.learner_ids(),
+            end_learners,
+            "End state learners inaccurate"
+        );
+        Ok(())
+    }
```

Comment:
Maybe it's impossible to happen in practice, but (similarly to a previous comment) it seems like the case where `begin_membership_change` is called twice before `finalize_membership_change` isn't represented in this file. If it's not possible, maybe an assertion in `begin_membership_change` is warrented; if it is possible, then tests.

Reply:
Other test ideas:

- finalize_membership_change called without a begin_membership_change
- finalize_membership_change called twice without intervening begin_membership_change

Reply:
Do `ProgressSet::voters_mut/learners_mut/iter_mut` need to be `pub`?  Seem like easy methods with which to break internal consistency. I don't see uses of voters_mut/learners_mut outside the mod, not sure about iter_mut.

With those mutable accessors, this module kinda wants a lot more tests of weird cases. Without those accessors there aren't as many weird cases.

Likewise with the `insert/remove` methods. If those can be called at any time, then it seems like there's a lot of mayhem one could cause, and more test cases to write. If those were eliminated and handled purely in the constructor or themselves via membership changes then there's fewer cases to think about.

Reply:
They don't need to be pub, but I think we should be cautious not to create too many API changes in this PR. It's more appropriate for a new PR.